### PR TITLE
Update waf version to 1.15.1

### PR DESCRIPTION
**Pull request by hoolioh**
### What does this PR do?

This PR updates the libddwaf version to 1.15.1. It contains breaking API changes:
* Add a new parameter to handle the whole payload: persistent and ephemeral addresses.
* Change function 'ddwaf_required_addresses' to 'ddwaf_known_addresses'.

[APPSEC-12207]

[APPSEC-12283]: https://datadoghq.atlassian.net/browse/APPSEC-12207

[APPSEC-12207]: https://datadoghq.atlassian.net/browse/APPSEC-12207?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -179,7 +180,7 @@ void DDWAF::update_required_addresses(const Napi::CallbackInfo& info) {
   Napi::Env env = info.Env();
 
   uint32_t size = 0;
-  const char* const* required_addresses = ddwaf_required_addresses(this->_handle, &size);
+  const char* const* required_addresses = ddwaf_known_addresses(this->_handle, &size);
```
Comment (uurien) [inline src/main.cpp:183]: why are we maintaining the `required_addresses` variable name, if it is not called required_address anymore?

File: src/main.cpp [hunk 2]
```diff
@@ -242,29 +243,45 @@ Napi::Value DDWAFContext::run(const Napi::CallbackInfo& info) {
...
-    Napi::TypeError::New(env, "First argument must be an object").ThrowAsJavaScriptException();
+
+  if (!info[0].IsObject() && !info[1].IsObject()) {
+    Napi::TypeError::New(env, "One of persistent data or ephemeral data must be an object").ThrowAsJavaScriptException();
     return env.Null();
   }
-  if (!info[1].IsNumber()) {
-    Napi::TypeError::New(env, "Second argument must be a number").ThrowAsJavaScriptException();
+
+  if (!info[2].IsNumber()) {
```
Comment (uurien) [inline src/main.cpp:257]: What if `info.Length() == 2`? Isn't this going to be a segmentation fault?

File: index.d.ts
Comment (uurien) [inline index.d.ts:1]: I understand that we should also modify L32 of this file

```
    run(inputs: object, ephemeral: object, timeout: number): result
```

File: src/main.cpp [hunk 3]
```diff
@@ -242,29 +243,45 @@ Napi::Value DDWAFContext::run(const Napi::CallbackInfo& info) {
     Napi::Error::New(env, "Calling run on a disposed context").ThrowAsJavaScriptException();
     return env.Null();
   }
+
   if (info.Length() < 2) {  // inputs, timeout
```
Comment (uurien) [inline src/main.cpp:247]: Maybe, this should be `< 3`?

File: test/index.js [hunk 4]
```diff
@@ -361,11 +422,15 @@ describe('DDWAF', () => {
     const waf = new DDWAF(rules)
     const context = waf.createContext()
 
-    assert.throws(() => context.run(), new Error('Wrong number of arguments, expected 2'))
-    assert.throws(() => context.run('', TIMEOUT), new TypeError('First argument must be an object'))
-    const err = new TypeError('Second argument must be greater than 0')
-    assert.throws(() => context.run({ 'server.request.headers.no_cookies': 'value_attack' }, -1), err)
-    assert.throws(() => context.run({ 'server.request.headers.no_cookies': 'value_attack' }, 0), err)
+    assert.throws(() => context.run(), new Error('Wrong number of arguments, at least 2 expected'))
```
Comment (uurien) [inline test/index.js:425]:
```suggestion
    const wrongArgsError = new Error('Wrong number of arguments, at least 3 expected')
    assert.throws(() => context.run(), wrongArgsError)
    assert.throws(() => context.run({ 'server.request.headers.no_cookies': 'value_attack' }), wrongArgsError)
    assert.throws(() => context.run({ 'server.request.headers.no_cookies': 'value_attack' }, TIMEOUT), wrongArgsError)
```

File: src/main.cpp (hunk 1 above)
Comment (hoolioh) [reply src/main.cpp]: You're right.

File: src/main.cpp (hunk 2 above)
Comment (hoolioh) [reply src/main.cpp:260]: Yep, I overlooked that.

File: src/main.cpp [hunk 5]
```diff
@@ -242,29 +245,56 @@ Napi::Value DDWAFContext::run(const Napi::CallbackInfo& info) {
...
-    Napi::TypeError::New(env, "First argument must be an object").ThrowAsJavaScriptException();
+    Napi::TypeError::New(
+            env,
+            "Payload data must be an object")
+        .ThrowAsJavaScriptException();
     return env.Null();
   }
+
+  Napi::Value persistent = info[0].As<Napi::Object>().Get("persistent");
+  Napi::Value ephemeral = info[0].As<Napi::Object>().Get("ephemeral");
```
Comment (uurien) [inline src/main.cpp:263]: Isn't a casting an expensive operation? why not:
```suggestion
  Napi::Object payload = info[0].As<Napi::Object>()
  Napi::Value persistent = payload.Get("persistent");
  Napi::Value ephemeral = payload.Get("ephemeral");
```

**Review by uurien — COMMENTED** [review]
No tests with `ephemeral`?

File: src/main.cpp (hunk 5 above)
Comment (hoolioh) [reply src/main.cpp:263]: If I recall correctly that will call to `napi_coerce_to_object` function which will do a `reinterpret_cast`. That type of casts don't have impact in runtime but probably the way you proposed it's more idiomatic.

File: index.d.ts [hunk 6]
```diff
@@ -39,7 +49,8 @@ export class DDWAF {
     custom_rules?: diagnosticsInfo,
     exclusions?: diagnosticsInfo,
     rules_override?: diagnosticsInfo,
-    rules_data?: diagnosticsInfo
+    rules_data?: diagnosticsInfo,
+    processors?: diagnosticsInfo
   };
 
   readonly requiredAddresses: Set<string>;
```
Comment (uurien) [inline index.d.ts:56]:
```suggestion
  readonly knownAddresses: Set<string>;
```

File: test/index.js [hunk 7]
```diff
@@ -64,13 +75,13 @@ describe('DDWAF', () => {
   it('should collect an attack and cleanup everything', () => {
     const waf = new DDWAF(rules)
     const context = waf.createContext()
+    const payload = {
+      persistent: {
+        'server.request.headers.no_cookies': 'value_ATTack'
```
Comment (uurien) [inline test/index.js:80]: you have removed data from the payload, why?

```
      x: new Array(4096).fill('x').join(''),
      y: new Array(4097).fill('y').join(''),
      z: new Array(4097).fill('z')
```

Comment (hoolioh) [reply test/index.js]: I've removed it accidentally when I refactored it.

**Review by uurien — APPROVED** [review]

File: src/main.cpp [hunk 8]
```diff
@@ -242,29 +245,57 @@ Napi::Value DDWAFContext::run(const Napi::CallbackInfo& info) {
...
   if (timeout <= 0) {
-    Napi::TypeError::New(env, "Second argument must be greater than 0").ThrowAsJavaScriptException();
+    Napi::TypeError::New(env, "Timeout argument must be greater than 0").ThrowAsJavaScriptException();
     return env.Null();
   }
 
+  ddwaf_object *ddwafPersistent = nullptr;
+
+  if (persistent.IsObject()) {
+    ddwafPersistent = static_cast<ddwaf_object *>(alloca(sizeof(ddwaf_object)));
```
Comment (iunanua) [inline src/main.cpp:285]: Why the allocation? in other cases it's not needed, no?

Comment (hoolioh) [reply src/main.cpp:285]: That allocation occurs in the stack. `alloca` function allows you to allocate space in the stack conditionally so if the space is not needed the stack won't grow. Regarding the space allocated, it's automatically deallocated when the function ends due to the stack frame is freed.

**Review by iunanua — APPROVED** [review]

**Pull request merged**